### PR TITLE
#3480 Fix new ssh key adding issues

### PR DESCRIPTION
**Pull request by mohamme89d**
https://github.com/gogits/gogs/issues/3480
Added regular exp match (108) and remove training new line (111) that cause system to think its an ssh2 key.

tested with Windows OpenSSH and OpenSSH2 keys

### CODE REVIEW

**Comment by Unknwon** [comment]
I think what you did with _regular exp match_ is overkill, all we need is `content = strings.TrimRight(content, "\n")`..

**Comment by mohamme89d** [comment]
I did try few tools and sometimes the comments part come in a new line so this will cater of few edge cases.

File: models/ssh_key.go [hunk 1]
```diff
@@ -104,8 +105,10 @@ func extractTypeFromBase64Key(key string) (string, error) {
 
 // parseKeyString parses any key string in OpenSSH or SSH2 format to clean OpenSSH string (RFC4253).
 func parseKeyString(content string) (string, error) {
-	// Transform all legal line endings to a single "\n".
-	content = strings.NewReplacer("\r\n", "\n", "\r", "\n").Replace(content)
+	// Transform all legal line endings to a single "\n"
+	newlineExpression := regexp.MustCompile(`\r?\n`) // windows style optional with (\r?)
+	content = newlineExpression.ReplaceAllString(content, "\n")
```
Comment (bkcsoft) [inline models/ssh_key.go]: Why not just `content = strings.Replace(content, "\r\n", "\n")` ?

**Comment by Unknwon** [comment]
>I did try few tools and sometimes the comments part come in a new line so this will cater of few edge cases.

Any examples?

File: models/ssh_key.go (hunk 1 above)
Comment (mohamme89d) [inline models/ssh_key.go]: > content = strings.Replace(content, "\r\n", "\n")  
will only fix windows new lines

Comment (bkcsoft) [inline models/ssh_key.go]: well, `\r?\n` will only match `\r\n` and `\n`, not `\r` like the previous implementation... I think you're looking for `\r\n?` in that case

**Comment by mohamme89d** [comment]
Did try tools below to copy the ssh-key public content 

1. Git kraken 
2. Windows Notepad
3. Notepad++ 

all of then append a new line at the end (which i think this is how ssh-keygen getenrate ssh-key [ssh-keygen from git 2.9.x])
capture


Case 1 : 
On Windows we when the old expression ran it replaces '\r\n' to '\n\n' which is two new line and it throws the whole thing out of desired behavior 
`old replacer uses ("\r\n", "\n", "\r", "\n"), but not sure why go is actually making "\r" more prioretized than "\r\n"
by right it should find "\r\n" and first but it seems like it goes for "\r" first so we end up with ("\r\n" --> "\n\n")
`
fix for that was 
1- define a regular expression that matches both environemts 

> regexp.MustCompile(`\r?\n`) // windows style optional with (\r?) 

2- match against the regular expression and replace matches 

>   content = newlineExpression.ReplaceAllString(content, "\n") 

Case 2 :
On single line key will have an extra (Windows '\r\n' or Linux '\n') at the end to fix that we have 
> strings.TrimRight(content, "\n")
Which just remove any additional new line at the end which is useless in all cases

**Comment by Unknwon** [comment]
>Case 1 : 
On Windows we when the old expression ran it replaces '\r\n' to '\n\n' which is two new line and it throws the whole thing out of desired behavior 
old replacer uses ("\r\n", "\n", "\r", "\n"), but n

We can use two `strings.Replace()` to fix this.

**Comment by mohamme89d** [comment]
The key is not small (looking at scale) using two replaces will scan the key content twice and generate 2 new immutable strings, but if regular expression is used we scan it once then matches are replaced generation only 1 new immutable string.

File: models/ssh_key.go (hunk 1 above)
Comment (mohamme89d) [inline models/ssh_key.go]: That is ture, but we should not get \r only, if that happens it means the input end is not a user with an interface ... it is a script.

**Comment by Unknwon** [comment]
Key is not big enough to have such concern, I prefer more readable code.

**Comment by mohamme89d** [comment]
Cool, will apply the changes and push.

**Comment by KokaKiwi** [comment]
Hi, why not merging the PR?

**Comment by Unknwon** [comment]
Thanks!

**Pull request merged**